File: apps/sdkwork-rtc-flutter-mobile/packages/sdkwork_rtc_flutter_mobile_rtc/lib/src/services/media_session_service.dart
import 'package:sdkwork_rtc_app_sdk/sdkwork_rtc_app_sdk.dart'
    as generated;
import 'package:sdkwork_rtc_flutter_mobile_core/sdkwork_rtc_flutter_mobile_core.dart';

import '../models/media_session.dart';
import 'rtc_command_idempotency.dart';

class MediaSessionListParams {
  final int? page;
  final int? pageSize;
  final String? cursor;
  final String? search;
  final String? sort;

  const MediaSessionListParams({
    this.page,
    this.pageSize,
    this.cursor,
    this.search,
    this.sort,
  });
}

class MediaSessionListResult {
  final List<RtcMediaSession> items;
  final String? nextCursor;

  const MediaSessionListResult({
    required this.items,
    this.nextCursor,
  });
}

class MediaSessionService {
  final SdkworkAppClient _client;

  MediaSessionService(this._client);

  Future<MediaSessionListResult> list([MediaSessionListParams? params]) async {
    final response = await _client.rtcMediaSessions.list(
      params?.page,
      params?.pageSize,
      params?.cursor,
      params?.sort,
      params?.search,
    );
    final envelope = _typedResponseEnvelope(response);
    final page = appSdkEnvelopeListPageFromMap(
      envelope,
      'RTC media session list response missing items',
    );
    return MediaSessionListResult(
      items: page.items.map(RtcMediaSession.fromJson).toList(),
      nextCursor: page.nextCursor,
    );
  }

  Future<RtcMediaSession> get(String mediaSessionId) async {
    final response = await _client.rtcMediaSessions.retrieve(mediaSessionId);
    final envelope = _typedResponseEnvelope(response);
    final entity = appSdkEnvelopeEntityFromMap(
      envelope,
      'RTC media session not found: $mediaSessionId',
    );
    return RtcMediaSession.fromJson(entity);
  }

  Future<RtcMediaSession> create(
    RtcCreateMediaSessionRequest body, {
    String? idempotencyKey,
  }) async {
    final response = await _client.rtcMediaSessions.create(
      generated.RtcCreateMediaSessionRequest(
        roomId: body.roomId,
        mediaMode: body.mediaMode,
        providerProfileId: body.providerProfileId,
        provider: body.provider,
      ),
      idempotencyKey ??
          createRtcCommandIdempotencyKey('media-session-create'),
    );
    final envelope = _typedResponseEnvelope(response);
    final entity = appSdkEnvelopeEntityFromMap(
      envelope,
      'RTC media session was not created',
    );
    return RtcMediaSession.fromJson(entity);
  }

  Map<String, dynamic>? _typedResponseEnvelope(dynamic response) {
    if (response == null) {
      return null;
    }
    final dynamic value = response;
    final data = value.data;
    if (data == null) {
      return <String, dynamic>{
        if (value.code != null) 'code': value.code,
        if (value.requestId != null) 'requestId': value.requestId,
      };
    }
    final dataJson = data is Map<String, dynamic>
        ? data
        : data is Map
            ? Map<String, dynamic>.from(data)
            : (data.toJson is Map<String, dynamic> Function()
                ? (data.toJson as Map<String, dynamic> Function())()
                : null);
    if (dataJson == null) {
      return null;
    }
    if (dataJson.containsKey('item') || dataJson.containsKey('items')) {
      return <String, dynamic>{'code': 0, 'data': dataJson};
    }
    return <String, dynamic>{'code': 0, 'data': <String, dynamic>{'item': dataJson}};
  }
}

File: apps/sdkwork-rtc-h5/packages/sdkwork-rtc-h5-admin-core/src/services/mediaSessionService.ts
import type { AuthTokenManager } from "@sdkwork/sdk-common";

import type {
  MediaSessionListParams,
  MediaSessionListResponse,
  RtcMediaSession,
} from "../types/mediaSession";
import type { RtcMediaSessionCompletionRecord } from "../types/completionRecord";
import {
  resolveBackendRtcClient,
  type RtcBackendClientOptions,
  type RtcBackendClientSource,
} from "./backendClient";

export class MediaSessionService {
  private readonly client;

  constructor(
    baseUrlOrClient: RtcBackendClientSource,
    tokenManagerOrOptions?: AuthTokenManager | RtcBackendClientOptions,
  ) {
    this.client = resolveBackendRtcClient(baseUrlOrClient, tokenManagerOrOptions);
  }

  async list(params?: MediaSessionListParams): Promise<MediaSessionListResponse> {
    const response = await this.client.rtcMediaSessions.rtc.mediaSessions.list({
      page: params?.page,
      pageSize: params?.limit,
      cursor: params?.cursor,
      q: params?.search,
      sort: params?.sort,
      status: params?.status,
      ownerUserId: params?.ownerUserId,
      createdAfter: params?.createdAfter,
    });
    // The generated SDK unwraps the sdkwork-v3 envelope, so the response is
    // already the `data` payload ({ items, pageInfo }).
    const nextCursor = response.pageInfo?.nextCursor;
    return {
      items: response.items,
      nextCursor: nextCursor && nextCursor.length > 0 ? nextCursor : undefined,
    };
  }

  async get(id: string): Promise<RtcMediaSession> {
    const response = await this.client.rtcMediaSessions.rtc.mediaSessions.retrieve(id);
    if (!response) {
      throw new Error(`RTC media session not found: ${id}`);
    }
    return response;
  }

  async close(id: string): Promise<RtcMediaSession> {
    const response = await this.client.rtcMediaSessions.rtc.mediaSessions.close(id, {});
    return response;
  }

  async getCompletionRecord(id: string): Promise<RtcMediaSessionCompletionRecord> {
    const response = await this.client.rtcMediaSessions.rtc.mediaSessions.completionRecord.retrieve(
      id,
    );
    if (!response) {
      throw new Error(`RTC media session completion record not found: ${id}`);
    }
    return response;
  }
}
